### PR TITLE
NIFI-1107 - Integrate Multipart uploads into the PutS3Object processor.

### PR DESCRIPTION
**Pull request by jskora**
* Add nifi-ssl-context-service-nar to nifi-aws-nar pom.
* Add nifi-ssl-context-service-api to nifi-aws-processors pom.
* Add SSL context service and endpoint override properties to AbstractAWSProcessor to support non-Amazon S3-compatible endpoints.
* Added Multipart upload API support, multipart threshold and size properties, state tracking for multipart upload, and expanded flow file attributes to PutS3Object.
* Updated annotations and property list for FetchS3Object and DeleteS3Object.
* Added region to bucket in AbstractS3Test to minimize AWS propogation issues during testing.
* Expanded tests in TestPutS3Object to validate new multipart logic.

### CODE REVIEW

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java [hunk 1]
```diff
@@ -102,6 +177,94 @@ protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String
...
+        if (descriptor.equals(KEY)
+                || descriptor.equals(BUCKET)
+                || descriptor.equals(ENDPOINT_OVERRIDE)
+                || descriptor.equals(STORAGE_CLASS)
+                || descriptor.equals(REGION)) {
+            destroyState();
+        }
+    }
+
+    protected MultipartState getState(final String s3ObjectKey) throws IOException {
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java]: I believe these methods could be problematic if multiple threads are monkeying with the persistence file at the same time.

Comment (joewitt) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java]: properties can only be modified if the processor is stopped so it *should be safe* i believe.

Comment (trkurc) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java]: @joewitt  - I was talking about the *State methods not the onPropertyModified.

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java [hunk 2]
```diff
@@ -143,23 +316,223 @@ public void process(final InputStream rawIn) throws IOException {
...
+                                        .withInputStream(in)
+                                        .withPartNumber(part)
+                                        .withPartSize(thisPartSize);
+                                try {
+                                    UploadPartResult uploadPartResult = s3.uploadPart(uploadRequest);
+                                    currentState.addPartETag(uploadPartResult.getPartETag());
+                                    currentState.setFilePosition(currentState.getFilePosition() + thisPartSize);
+                                    try {
+                                        persistState(cacheKey, currentState);
+                                    } catch (Exception e) {
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java]: I'd add a comment here on why swallowing exceptions here is okay, as it is not immediately intuitive.

**Comment by trkurc** [comment]
My major concern is with the persistence file. I would think scrubbing the file periodically, removing older stale entries would be a must, otherwise the file will begin accumulating cruft as more and more files don't complete the multipart lifecycle.

**Comment by trkurc** [comment]
Another, major secondary concern is what to do with failed partial multipart puts and maybe having to do bucket cleanup.

**Comment by jskora** [comment]
I'm closing this pull request since the general S3 changes have been incorporated by @trkurc.  The remaining Multipart Upload functionality will be incorporated into a new pull request in conjunction with S3 bucket and local state cleanup functionality.

**Pull request closed without merging**